### ISSUE TITLE
Cache Bounds of List Watchers

### ISSUE THREAD

**Issue by cycomachead** [jmoenig/Snap--Build-Your-Own-Blocks#1076]
It would be helpful for list watchers to cache their bounds. If you're working on a big list, and you make the watcher larger, then when you use `set` on that variable, the size/position of the watcher should stay the same.

**Comment by jmoenig** [comment]
list watchers stay the same size as you mutate the list. *variable* watchers reset if you *assign* a new value to the variable. If you use functional programming list watchers aren't of much use until you have the result anyway, because you get *a new list* with every operation. To me this is very important. If, on the other hand you work imperatively, list watchers already visualize mutation by keeping their dimensions.

In other words: I believe that getting a *new list watcher* whenever you create a *new list* is pedagogically important. See, even you think that it's somehow the "same" list watcher, implying that it's somehow the "same" list, but that's totally a misconception.

**Comment by brianharvey** [comment]
Perhaps it'll help when you size the watcher to the size of the list, when a variable is assigned a new value, instead of making it tiny -- you did say you're going to do that, right?

**Comment by jmoenig** [comment]
I *did* do that, but not for the stage, because on the stage you often want to place and dimension your watchers explicitly. The issue is about *mutation* vs. *assignment*. If you want the *watcher* to stay the same you're basically calling for a *variable type declaration*. I hate how static typing keeps coming up.

**Comment by cycomachead** [comment]
> If you use functional programming list watchers aren't of much use until you have the result anyway, because you get a new list with every operation.

Why? This might be the case while running a particular specifc operation, but this _hardly_ the case while exploring operations or demoing things. When presenting you often do this with non-functional list ops. 

Other use case -- this one I should have submitted long ago.
Often times you'll want to use a scores list for a project, or a play names list. When you reopen a project, the list resets. This leads students to do really awkward things like start off a project with a `delete all` rather than a `set`. 

> In other words: I believe that getting a new list watcher whenever you create a new list is pedagogically important. See, even you think that it's somehow the "same" list watcher, implying that it's somehow the "same" list, but that's totally a misconception.

I honestly don't see why this is a decision about pedagogy or teaching? I don't think most people think of or use watchers in a way that relates to identity of variables. But I also think that based on current practice, what you say is somewhat wrong. A _new_ watcher is adding to a left column of the stage, below any watchers that already exist. It also starts out with a "normal" view. Resetting a variable does _not_ change the watcher position nor the watcher display type.

**Comment by jmoenig** [comment]
If you use non-functional list ops you *mutate* the list and the *watcher* does stay the same, including its dimensions! Only if you assign the variable *again* to a CONS you get a new watcher because you get a new list. That's the *point* of functional programming, right?

You can totally demo intermediate steps, e.g. the different stages of the BJC *acronym* example, and I've changed it so you always see the whole list in speech bubbles or result balloons in the scripting area. Isn't that *exactly* how you want to demo functional programming?

**Comment by brianharvey** [comment]
> I did do that, but not for the stage, because on the stage you often want to place and dimension your watchers explicitly. 

When you combine what you said to me (what I quoted) with what you said to Michael you are really putting functional programmers in an impossible situation.  I would *love* to place and dimension my watchers explicitly, but you won't let me; you shrink them to unreadable if I create a new list (or a new anything; this thing about functional vs imperative list construction is really a red herring -- think about assigning a procedure to a variable instead).

You could make everyone happy without pushing your button about static typing by just honoring the user dragging out a watcher; make that permanent until the user changes it.  Or, if that's too much, put a little pushpin icon button next to the resize handle and if the user pushes it you leave the size alone.

**Comment by jmoenig** [comment]
I'm fuming mad about this debate. It's static typing all over, sooner or later educators want bondage and discipline. Tell me one demo you want to give to students that you can't!

**Comment by brianharvey** [comment]
I think you're so fuming mad you've stopped reading carefully.  My position can't be characterized as static typing.  I'll give you that that's a possible reading of an implication of Michael's position, although even that's arguable.  Certainly it's not Michael's *intention* and you're being unfair to him by characterizing it that way.  What his position and mine have in common is that neither of us is thinking about a teacher in front of the class; we're thinking about a plain old user, maybe not even *in* a class, who's trying to use a watcher to watch something and keeps having to drag the damn watcher out.  He's not thinking about types one way or the other, and he's not even thinking about functional vs imperative style.  He's just trying to get his project done!

And you know me well enough to know how I feel about b&d languages.

**Comment by cycomachead** [comment]
> If you use non-functional list ops you mutate the list and the watcher does stay the same, including its dimensions! Only if you assign the variable again to a CONS you get a new watcher because you get a new list. That's the point of functional programming, right?
> 
> You can totally demo intermediate steps, e.g. the different stages of the BJC acronym example, and I've changed it so you always see the whole list in speech bubbles or result balloons in the scripting area. Isn't that exactly how you want to demo functional programming?

First point: Even when doing doing non-functional programming you, especially while exploring, you commonly will use something like `set (my list) to [A, B, C]`. That is what resets the watcher. 

When doing purely functional programming, just clicking on reports works pretty well, but watchers let you keep data around. You could be comparing 2 different functional operations and want to use the stage. 

> Tell me one demo you want to give to students that you can't!

_Can't_ is almost always the wrong idea with these requests. Certainly, we _can_ demo lots of projects, or explore lists, or do lots of things in Snap<em>!</em>. The problem is that some are frustating. When I see people repeat the same action 5 times over because Snap<em>!</em> is resetting a watcher, then we should make that experience better. Or find a better way for them to accomplish the same task that is easy.
